File: source/Managed/Math.cs
// (c) gfoidl, all rights reserved

using System;
using System.Runtime.InteropServices;

namespace Managed;

public static class Math
{
    public static int Add(int a, int b) => Native.Add(a, b);
    //---------------------------------------------------------------------------------------------
    public static unsafe int Sum(ReadOnlySpan<int> data)
    {
        if (data.IsEmpty)
        {
            return 0;
        }

        fixed (int* ptr = &MemoryMarshal.GetReference(data))
        {
            return Native.Sum(ptr, data.Length);
        }
    }
    //---------------------------------------------------------------------------------------------
    public static string GetNativeVersion()     => Native.GetVersion();
    public static string GetNativeBuildConfig() => Native.BuildConfig();
}

File: source/Native/math.cpp
#include "math.h"
//-----------------------------------------------------------------------------
int add(const int a, const int b)
{
    return a + b;
}
//-----------------------------------------------------------------------------
int sum(const int* arr, const int len)
{
    int res = 0;

    for (int i = 0; i < len; ++i)
    {
        res += arr[i];
    }

    return res;
}
